### PR TITLE
Implement bulk data upload and delete endpoints

### LINKED ISSUE #210: Implement Bulk Data Upload and Delete endpoints
## Summary

The gateway already implements **bulk data read** (listing categories, listing items, downloading items). This issue adds the missing write and delete operations:

- **Upload (POST):** Upload binary data to a bulk data category via multipart form-data
- **Delete (DELETE):** Remove a specific bulk data item

These complete the bulk data CRUD cycle.

---

### Existing Endpoints (already implemented)

For reference, the following bulk data read endpoints already exist:

| Method | Path | Description |
|--------|------|-------------|
| `GET` | `/{entity}/bulk-data` | List bulk data categories |
| `GET` | `/{entity}/bulk-data/{category}` | List items in a category |
| `GET` | `/{entity}/bulk-data/{category}/{item-id}` | Download a bulk data item |
| `GET` | `/{entity}/faults/{fault-id}/bulk-data/{category}/{item-id}` | Download fault-associated bulk data |

---

### 1. `POST /api/v1/{entity-path}/bulk-data/{category}`

Upload a new bulk data item to a specific category.

**Applies to entity types:** Components, Apps

**Path parameters:**

| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `{entity-path}` | URL segment | Yes | e.g., `apps/temp_sensor` |
| `{category}` | `string` | Yes | Bulk data category (e.g., `calibration`, `firmware`, `comlogs`, `snapshots`) |

**Content-Type:** `multipart/form-data`

**Multipart fields:**

| Field | Type | Required | Description |
|-------|------|----------|-------------|
| `file` | binary | Yes | The file to upload |
| `description` | `string` | No | Human-readable description |
| `metadata` | `application/json` | No | Structured metadata as JSON |

**Example request (conceptual):**

```
POST /api/v1/apps/temp_sensor/bulk-data/calibration HTTP/1.1
Content-Type: multipart/form-data; boundary=----FormBoundary

------FormBoundary
Content-Disposition: form-data; name="file"; filename="calibration_2026.dat"
Content-Type: application/octet-stream

<binary data>
------FormBoundary
Content-Disposition: form-data; name="description"

Temperature sensor calibration data from 2026-02-14
------FormBoundary
Content-Disposition: form-data; name="metadata"
Content-Type: application/json

{"sensor_type": "thermocouple", "calibration_date": "2026-02-14"}
------FormBoundary--
```

**Response `201 Created`:**

```json
{
  "id": "cal_2026_02_14_001",
  "category": "calibration",
  "name": "calibration_2026.dat",
  "mime_type": "application/octet-stream",
  "size": 102400,
  "created": "2026-02-14T10:30:00Z",
  "description": "Temperature sensor calibration data from 2026-02-14",
  "metadata": {
    "sensor_type": "thermocouple",
    "calibration_date": "2026-02-14"
  }
}
```

This is a `BulkDataDescriptor` object:

| Field | Type | Required | Description |
|-------|------|----------|-------------|
| `id` | `string` | Yes | Server-generated unique identifier |
| `category` | `string` | Yes | Category this item belongs to |
| `name` | `string` | Yes | Original uploaded filename |
| `mime_type` | `string` | Yes | MIME type of the uploaded file |
| `size` | `integer` | Yes | File size in bytes |
| `created` | `string` (ISO 8601) | Yes | Upload timestamp |
| `description` | `string` | No | Human-readable description |
| `metadata` | `object` | No | Structured metadata |

**Location header:** `/api/v1/apps/temp_sensor/bulk-data/calibration/cal_2026_02_14_001`

**Error responses:**

| Status | Error Code | When |
|--------|-----------|------|
| `400` | `invalid-request` | No file part in multipart, or empty file |
| `400` | `invalid-parameter` | Unknown category |
| `404` | `entity-not-found` | Entity doesn't exist |
| `413` | `invalid-request` | File exceeds maximum upload size (configurable) |

---

### 2. `DELETE /api/v1/{entity-path}/bulk-data/{category}/{item-id}`

Delete a specific bulk data item.

**Path parameters:**

| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `{entity-path}` | URL segment | Yes | e.g., `apps/temp_sensor` |
| `{category}` | `string` | Yes | Bulk data category |
| `{item-id}` | `string` | Yes | Item identifier |

**Response `204 No Content`**

**Error responses:**

| Status | Error Code | When |
|--------|-----------|------|
| `404` | `entity-not-found` | Entity doesn't exist |
| `404` | `resource-not-found` | Category or item doesn't exist |

---

### Multipart parsing with cpp-httplib

cpp-httplib has built-in support for multipart form data:

```cpp
void BulkDataHandlers::handle_upload(const httplib::Request& req, httplib::Response& res) {
  auto entity_id = req.matches[1].str();
  auto category = req.matches[2].str();

  auto entity = ctx_.validate_entity_for_route(req, res, entity_id);
  if (!entity) return;

  // Check if the request has the file part
  if (!req.has_file("file")) {
    ctx_.send_error(res, 400, ERR_INVALID_REQUEST, "Missing 'file' part in multipart upload");
    return;
  }

  auto file = req.get_file_value("file");
  // file.filename - original filename
  // file.content_type - MIME type
  // file.content - binary data

  // Optional fields
  std::string description;
  if (req.has_file("description")) {
    description = req.get_file_value("description").content;
  }
}
```

### Storage

- Store uploaded files on disk in `{bulk_data_dir}/{entity_id}/{category}/{item_id}/`
- Maintain metadata (BulkDataDescriptor) in memory or a sidecar JSON file
- The `{bulk_data_dir}` should be configurable via gateway parameters

### File management

- **ID generation:** Use a combination of category + timestamp + counter for uniqueness
- **Size limits:** Configurable max upload size (default: 100MB). cpp-httplib can set `server.set_payload_max_length()`
- **Disk cleanup:** Consider optional auto-cleanup when disk usage exceeds threshold

### Extending existing handler

The existing `bulkdata_handlers.cpp` already handles GET requests. Add POST and DELETE handlers to the same class:

```cpp
// In BulkDataHandlers class
void handle_upload(const httplib::Request& req, httplib::Response& res);
void handle_delete(const httplib::Request& req, httplib::Response& res);
```

### Route registration

Add to the existing bulk data routes:

```cpp
// Upload
srv->Post((api_path("/components") + R"(/([^/]+)/bulk-data/([^/]+)$)"), handler);
srv->Post((api_path("/apps") + R"(/([^/]+)/bulk-data/([^/]+)$)"), handler);

// Delete
srv->Delete((api_path("/components") + R"(/([^/]+)/bulk-data/([^/]+)/([^/]+)$)"), handler);
srv->Delete((api_path("/apps") + R"(/([^/]+)/bulk-data/([^/]+)/([^/]+)$)"), handler);
```

### Tests

- Unit test: upload file → 201 with correct BulkDataDescriptor
- Unit test: upload without file part → 400
- Unit test: upload to unknown category → 400
- Unit test: upload exceeding size limit → 413
- Unit test: list after upload → includes new item
- Unit test: download after upload → returns original content
- Unit test: delete existing item → 204
- Unit test: delete non-existent item → 404
- Unit test: list after delete → item removed

### PR DESCRIPTION
**Pull request by bburda**
# Pull Request

## Summary

Implement bulk data **upload** (`POST`) and **delete** (`DELETE`) endpoints for the SOVD bulk-data resource collection, complementing the existing read-only `GET` handlers.

**Key changes:**
- `BulkDataStore` - new filesystem-based CRUD engine with configurable storage dir, upload size limit, and category allowlist
- `POST /api/v1/{entity-path}/bulk-data/{category}` - multipart/form-data upload with optional description + metadata
- `DELETE /api/v1/{entity-path}/bulk-data/{category}/{id}`- remove uploaded items (rosbags remain read-only via FaultManager)
- Extended existing `GET` list/download handlers to serve both rosbag and non-rosbag categories
- Refactored `stream_file_to_response` into a generic utility (extracted from rosbag-specific logic)
- Added `validate_collection_access(BULK_DATA)` to all existing bulk data handlers (was missing)
- `BulkDataStore` injected via `HandlerContext`, configured from `gateway_params.yaml`

## Issue

Link the related issue (required):

- closes #210 

---

## Type

- [ ] Bug fix
- [x] New feature or tests
- [ ] Breaking change
- [ ] Documentation only

---

## Testing

- **Unit tests:** `test_bulk_data_store.cpp`  - store/remove/list/validate, size limits, edge cases
- **Unit tests:** `test_bulkdata_handlers.cpp` - descriptor JSON serialization, conditional fields
- **Integration tests:** `test_bulkdata_upload.test.py` - full POST/DELETE cycle via HTTP against running gateway
- **Integration tests:** updated `test_integration.test.py` - fixed `test_122`/`test_130` to respect SOVD (areas don't support resource collections)
- All bulk data unit tests pass
- Build clean with no warnings (`colcon build --packages-select ros2_medkit_gateway`)

```bash
# Verify locally
colcon build --packages-select ros2_medkit_gateway
cd build/ros2_medkit_gateway && ctest -R "test_bulk_data_store|test_bulkdata_handlers" --output-on-failure
```
---

## Checklist

- [x] Breaking changes are clearly described (and announced in docs / changelog if needed)
- [x] Tests were added or updated if needed
- [x] Docs were updated if behavior or public API changed

### CODE REVIEW

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+                                                                     const std::string & category) const {
+  std::vector<ItemDescriptor> items;
+
+  auto cat_dir = std::filesystem::path(storage_dir_) / entity_id / category;
+  if (!std::filesystem::is_directory(cat_dir)) {
+    return items;
+  }
+
+  std::lock_guard<std::mutex> lock(mutex_);
+  for (const auto & entry : std::filesystem::directory_iterator(cat_dir)) {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp:290]: Moved `is_directory()` check inside `mutex_` lock

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+std::optional<BulkDataStore::ItemDescriptor> BulkDataStore::get_item(const std::string & entity_id,
+                                                                     const std::string & category,
+                                                                     const std::string & item_id) const {
+  auto dir = item_dir(entity_id, category, item_id);
+  if (!std::filesystem::is_directory(dir)) {
+    return std::nullopt;
+  }
+
+  std::lock_guard<std::mutex> lock(mutex_);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp]: Moved `is_directory()` check inside `mutex_` lock

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+std::optional<std::string> BulkDataStore::get_file_path(const std::string & entity_id, const std::string & category,
+                                                        const std::string & item_id) const {
+  auto dir = item_dir(entity_id, category, item_id);
+  if (!std::filesystem::is_directory(dir)) {
+    return std::nullopt;
+  }
+
+  std::lock_guard<std::mutex> lock(mutex_);
+  // Find the payload file (not descriptor.json and not .data.tmp)
+  for (const auto & entry : std::filesystem::directory_iterator(dir)) {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp:325]: Moved `is_directory()` check inside `mutex_` lock

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    std::filesystem::rename(tmp_path, final_path, ec);
+    if (ec) {
+      // Fallback: try copy + remove (cross-device rename)
+      std::filesystem::copy_file(tmp_path, final_path, ec);
+      if (ec) {
+        return tl::unexpected("Failed to finalize file: " + ec.message());
+      }
+      std::filesystem::remove(tmp_path, ec);
+    }
+    write_descriptor(dir, desc);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp]: Added `std::filesystem::remove(tmp_path)` cleanup on rename/copy failure

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+  if (auto err = validate_path_component(item_id, "item_id"); !err) {
+    return tl::unexpected(err.error());
+  }
+
+  auto dir = item_dir(entity_id, category, item_id);
+  if (!std::filesystem::is_directory(dir)) {
+    return tl::unexpected("Bulk-data item not found: " + item_id);
+  }
+
+  std::lock_guard<std::mutex> lock(mutex_);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp]: Moved `is_directory()` check inside `mutex_` lock

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  j["size"] = desc.size;
+  j["created"] = desc.created;
+  j["description"] = desc.description;
+  if (!desc.metadata.empty()) {
+    j["metadata"] = desc.metadata;
+  }
+
+  auto path = dir / "descriptor.json";
+  std::ofstream ofs(path);
+  ofs << j.dump(2);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp:114]: Changed return type to `bool`, added `ofs.flush()` + `ofs.good()` check; `store()` now cleans up the item dir on descriptor write failure

File: src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp [hunk 7]
```diff
@@ -148,60 +311,161 @@ void BulkDataHandlers::handle_download(const httplib::Request & req, httplib::Re
...
+      return;
+    }
+
+    // Get descriptor for filename and MIME type
+    auto item = store->get_item(entity_info->entity_id, category, bulk_data_id);
+    std::string filename = item ? item->name : bulk_data_id;
+    std::string mimetype = item ? item->mime_type : "application/octet-stream";
+
+    // Content-Disposition with original filename
+    res.set_header("Content-Disposition", "attachment; filename=\"" + filename + "\"");
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp]: `filename` here comes from `item->name` which is the original upload filename. If someone uploads a file named `foo"bar.bin`, this produces a malformed Content-Disposition header: `attachment; filename="foo"bar.bin"`.

We should sanitize or escape quotes in the filename. Simplest fix:
```cpp
// Strip quotes from filename for Content-Disposition safety
std::string safe_name = filename;
std::replace(safe_name.begin(), safe_name.end(), '"', '_');
res.set_header("Content-Disposition", "attachment; filename=\"" + safe_name + "\"");
```
Alternatively, RFC 5987 encoding (`filename*=UTF-8''...`) handles this properly but is more complex.

File: src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp [hunk 8]
```diff
@@ -65,77 +80,225 @@ void BulkDataHandlers::handle_list_descriptors(const httplib::Request & req, htt
...
+  if (req.has_file("metadata")) {
+    const auto & meta_str = req.get_file_value("metadata").content;
+    if (!meta_str.empty()) {
+      auto parsed = nlohmann::json::parse(meta_str, nullptr, false);
+      if (parsed.is_discarded()) {
+        HandlerContext::send_error(res, httplib::StatusCode::BadRequest_400, ERR_INVALID_PARAMETER,
+                                   "Invalid JSON in 'metadata' field");
+        return;
+      }
+      metadata = std::move(parsed);
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp:275]: This accepts any valid JSON: arrays, strings, numbers, booleans. If someone sends `"metadata": "[1,2,3]"`, it gets stored as-is and shows up as `"x-medkit": [1,2,3]` in the API response. The descriptor schema expects an object here.

Suggestion - add a type check after the parse:
```cpp
if (!parsed.is_object()) {
    HandlerContext::send_error(res, httplib::StatusCode::BadRequest_400, ERR_INVALID_PARAMETER,
                               "metadata must be a JSON object");
    return;
}
```

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+// --- Validation ---
+
+tl::expected<void, std::string> BulkDataStore::validate_path_component(const std::string & value,
+                                                                       const std::string & name) {
+  if (value.empty()) {
+    return tl::unexpected(name + " cannot be empty");
+  }
+  // Reject path traversal and unsafe characters
+  if (value.find("..") != std::string::npos) {
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/bulk_data_store.cpp:47]: A single `"."` passes validation - `"."` as entity_id would resolve `{storage_dir}/./category/item_id/` to `{storage_dir}/category/item_id/`, causing a path collision across entities.

In practice this can't happen because entity_ids go through `validate_entity_for_route()` first, but for defense-in-depth:
```cpp
if (value == ".") {
    return tl::unexpected(name + " cannot be '.'");
}
```

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+
+  auto dir = item_dir(entity_id, category, item_id);
+
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (!std::filesystem::is_directory(dir)) {
+    return tl::unexpected("Bulk-data item not found: " + item_id);
+  }
+
+  std::error_code ec;
+  std::filesystem::remove_all(dir, ec);
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/bulk_data_store.cpp:269]: The lock is held for the entire `remove_all()` which blocks all other store operations (list, get, store finalization, other removes) while the directory tree is being deleted.

For our use case this is probably fine - deletes are rare and directories are small (1 file + descriptor.json). Just flagging it in case we ever store larger directory trees. If it becomes an issue, we could rename the dir to a `.deleting_` prefix under lock and then remove outside the lock.

File: src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp (hunk 7 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp]: Good catch! Added `std::replace` to strip quotes from filenames in both Content-Disposition locations (rosbag and store paths). Keeping it simple with quote-to-underscore replacement for now. RFC 5987 encoding would be overkill at this stage.

File: src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp (hunk 8 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/bulkdata_handlers.cpp:275]: Added suggestion. Thanks!

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp (hunk 9 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp:47]: Added suggestion!

File: src/ros2_medkit_gateway/src/bulk_data_store.cpp (hunk 10 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/bulk_data_store.cpp:269]: Added a NOTE comment documenting the rename-then-delete-outside-lock optimization for if/when this becomes a bottleneck.
